### ISSUE TITLE
dvc push: fails with Google Cloud Storage when remote URL is set to the bucket's root directory

### ISSUE THREAD

**Issue by m7rcelo** [iterative/dvc#6231]
# Bug Report

## Description

dvc push gives the error: 

> ERROR: unexpected error - 'name'

 when trying to push files to the default remote in a newly created project. Changing the remote's URL from `gs://mybucket` to `gs://mybucket/subfolder` fixed it.

### Reproduce

1. dvc init
2. Copy dataset.zip to the directory
3. dvc add dataset.zip
4. dvc remote add -d myremote gs://mybucket
5. dvc push

### Expected

Push should be successful.

### Environment information

Ubuntu with snap dvc installation. Also happened with pip installation.

**Additional Information (if any):**

I don't think this should be the expected behavior, because, if a bucket was created exclusively to host the dvc files for the project, it would be normal to upload the results of the push command to the bucket's root directory. If this is expected behavior, at least someone googling the error will now find a result.

**Comment by cverluise** [comment]
Hello, had the same issue. 

Seems that a trailing `/` (e.g.  `gs://mybucket/`) is now required. at least it fixed the issue for me. 

Hope it can help

**Comment by pmrowla** [comment]
I think this is a regression, @isidentical we are still supposed default to using the root `/` even if it's not included in the URL, right?

**Comment by isidentical** [comment]
> I think this is a regression, @isidentical we are still supposed default to using the root / even if it's not included in the URL, right?

Right. I'll submit a patch for this

**Comment by cverluise** [comment]
Thanks @isidentical  and @pmrowla !

Also, if you can make the Error message a bit more explicit, that would help a lot IMHO.

Cheers!

**Comment by m7rcelo** [comment]
Thank you!

**Comment by isidentical** [comment]
I can't reproduce it on the master;
```
(.venv38) (Python 3.8.5+) [ 11:57ÖÖ ]  [ <hey>@<hey>:~/temp_remotes ]
 $ mkdir gcp_remote
(.venv38) (Python 3.8.5+) [ 11:57ÖÖ ]  [ <hey>@<hey>:~/temp_remotes ]
 $ cd gcp_remote
(.venv38) (Python 3.8.5+) [ 11:57ÖÖ ]  [ <hey>@<hey>:~/temp_remotes/gcp_remote ]
 $ git init && dvc init  
Initialized empty Git repository in /home/isidentical/temp_remotes/gcp_remote/.git/
Initialized DVC repository.

You can now commit the changes to git.

+---------------------------------------------------------------------+
|                                                                     |
|        DVC has enabled anonymous aggregate usage analytics.         |
|     Read the analytics documentation (and how to opt-out) here:     |
|             <https://dvc.org/doc/user-guide/analytics>              |
|                                                                     |
+---------------------------------------------------------------------+

What's next?
------------
- Check out the documentation: <https://dvc.org/doc>
- Get help and share ideas: <https://dvc.org/chat>
- Star us on GitHub: <https://github.com/iterative/dvc>
(.venv38) (Python 3.8.5+) [ 11:57ÖÖ ]  [ <hey>@<hey>:~/temp_remotes/gcp_remote(master✗) ]
 $ dvc remote add -d myremote gs://dvc-test
Setting 'myremote' as a default remote.
(.venv38) (Python 3.8.5+) [ 11:57ÖÖ ]  [ <hey>@<hey>:~/temp_remotes/gcp_remote(master✗) ]
 $ echo 'yy' > yy
(.venv38) (Python 3.8.5+) [ 11:58ÖÖ ]  [ <hey>@<hey>:~/temp_remotes/gcp_remote(master✗) ]
 $ dvc add yy
100% Add|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|1/1 [00:00,  3.56file/s]
                                                                                                                                                                                  
To track the changes with git, run:

        git add yy.dvc .gitignore
(.venv38) (Python 3.8.5+) [ 11:58ÖÖ ]  [ <hey>@<hey>:~/temp_remotes/gcp_remote(master✗) ]
 $ dvc push yy
1 file pushed                                                                                                                                                                     
```

@m7rcelo @cverluise Would you mind supplying me `dvc doctor` and `dvc push -v`'s output?

**Comment by m7rcelo** [comment]
Hello @isidentical,

`dvc doctor`:
```
DVC version: 2.4.1 (snap)
---------------------------------
Platform: Python 3.6.9 on Linux-5.4.0-77-generic-x86_64-with-Ubuntu-20.04-focal
Supports: azure, gdrive, gs, http, https, s3, ssh
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/mapper/sda_crypt
Caches: local
Remotes: gs
Workspace directory: ext4 on /dev/mapper/sda_crypt
Repo: dvc, git
```

`dvc push -v`:
```
2021-06-29 11:55:41,188 DEBUG: Preparing to upload data to 'gs://my-bucket'
2021-06-29 11:55:41,189 DEBUG: Preparing to collect status from gs://my-bucket
2021-06-29 11:55:41,223 DEBUG: Collecting information from local cache...
2021-06-29 11:55:41,325 DEBUG: Collecting information from remote cache...      
2021-06-29 11:55:41,336 DEBUG: Matched '0' indexed hashes
2021-06-29 11:55:42,980 DEBUG: Estimated remote size: 4096 files                
2021-06-29 11:55:42,981 DEBUG: Querying '17031' hashes via traverse             
2021-06-29 11:55:43,683 ERROR: unexpected error - 'name'                        
------------------------------------------------------------
Traceback (most recent call last):
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/main.py", line 55, in main
    ret = cmd.do_run()
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/command/base.py", line 50, in do_run
    return self.run()
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/command/data_sync.py", line 67, in run
    glob=self.args.glob,
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/repo/__init__.py", line 51, in wrapper
    return f(repo, *args, **kwargs)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/repo/push.py", line 44, in push
    pushed += self.cloud.push(objs, jobs, remote=remote)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/data_cloud.py", line 83, in push
    show_checksums=show_checksums,
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/remote/base.py", line 57, in wrapper
    return f(obj, *args, **kwargs)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/remote/base.py", line 499, in push
    download=False,
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/remote/base.py", line 356, in _process
    download=download,
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/remote/base.py", line 196, in _status
    md5s, jobs=jobs, name=str(self.odb.path_info)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/remote/base.py", line 145, in hashes_exist
    return indexed_hashes + self.odb.hashes_exist(list(hashes), **kwargs)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/objects/db/base.py", line 468, in hashes_exist
    self.list_hashes_traverse(remote_size, remote_hashes, jobs, name)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/objects/db/base.py", line 312, in list_hashes_traverse
    yield from itertools.chain.from_iterable(in_remote)
  File "/snap/dvc/1306/usr/lib/python3.6/concurrent/futures/_base.py", line 586, in result_iterator
    yield fs.pop().result()
  File "/snap/dvc/1306/usr/lib/python3.6/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/snap/dvc/1306/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/snap/dvc/1306/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/objects/db/base.py", line 304, in list_with_update
    prefix=prefix, progress_callback=pbar.update
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/objects/db/base.py", line 193, in list_hashes
    for path in self._list_paths(prefix, progress_callback):
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/objects/db/base.py", line 173, in _list_paths
    for file_info in self.fs.walk_files(path_info, prefix=prefix):
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/fs/fsspec_wrapper.py", line 109, in walk_files
    for file in self.find(path_info, **kwargs):
  File "/snap/dvc/1306/lib/python3.6/site-packages/dvc/fs/fsspec_wrapper.py", line 179, in find
    path, detail=detail, prefix=path_info.parts[-1]
  File "/snap/dvc/1306/lib/python3.6/site-packages/fsspec/asyn.py", line 87, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "/snap/dvc/1306/lib/python3.6/site-packages/fsspec/asyn.py", line 68, in sync
    raise result[0]
  File "/snap/dvc/1306/lib/python3.6/site-packages/fsspec/asyn.py", line 24, in _runner
    result[0] = await coro
  File "/snap/dvc/1306/lib/python3.6/site-packages/gcsfs/core.py", line 934, in _find
    path,
  File "/snap/dvc/1306/lib/python3.6/site-packages/gcsfs/core.py", line 401, in _get_object
    return self._process_object(bucket, res)
  File "/snap/dvc/1306/lib/python3.6/site-packages/gcsfs/core.py", line 361, in _process_object
    result["name"] = posixpath.join(bucket, object_metadata["name"])
KeyError: 'name'
------------------------------------------------------------
2021-06-29 11:55:46,272 DEBUG: Version info for developers:
DVC version: 2.4.1 (snap)
---------------------------------
Platform: Python 3.6.9 on Linux-5.4.0-77-generic-x86_64-with-Ubuntu-20.04-focal
Supports: azure, gdrive, gs, http, https, s3, ssh
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/mapper/sda_crypt
Caches: local
Remotes: gs
Workspace directory: ext4 on /dev/mapper/sda_crypt
Repo: dvc, git

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
2021-06-29 11:55:46,277 DEBUG: Analytics is enabled.
2021-06-29 11:55:46,368 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmpcqlmcu07']'
2021-06-29 11:55:46,370 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmpcqlmcu07']'
```

**Comment by m7rcelo** [comment]
@isidentical

**Comment by isidentical** [comment]
@m7rcelo can you try the latest version of DVC and report back whether it works or not?

**Comment by m7rcelo** [comment]
@isidentical This issue appears to be fixed on DVC 2.5.0. I was able to push to the root directory of the remote bucket with no problems. Thank you!
